FIVETRAN_CODING_PRINCIPLES.md

```diff
@@ -0,0 +1,192 @@
+4. Feed the project to cursor agent `connector_sdk_readme` to generate a consistent `README.md`.
+5. Submit the PR.
+
+---
+
+## 4. Example connector guidelines
+
+### 4.1 Template connector
+
+A reference implementation is available here:  
+🔗 [Fivetran Connector SDK Template Example](https://github.com/fivetran/fivetran_connector_sdk/tree/main/template_example_connector)
+
+This includes:
+- A `connector.py` template.
+- A `README.md` template illustrating recommended structure and conventions.
+
+This is a great place to start to learn our coding practices through templated examples.
+
+
+### 4.2 General best practices
+
+- Import only necessary modules and libraries.
+- Use clear, consistent, descriptive names for functions and variables.
+- Use UPPERCASE_WITH_UNDERSCORES for constants (e.g., `CHECKPOINT_INTERVAL`).
+- Keep constants private by default (e.g., `__CHECKPOINT_INTERVAL`).
+- Add docstrings explaining each function’s purpose.
+- Add comments for complex logic and user customization points.
+- Split large functions into smaller helper functions for readability.
+- Use logging judiciously — avoid excessive or redundant logs.
+- Implement error handling that targets specific exceptions.
+- Add retry logic (with exponential backoff) for API requests.
+- Define a complete data model in the `schema()` function, with primary keys and data types.
+- Avoid loading all data into memory — use pagination or streaming.
+- Document pagination, upsert/update/delete, and checkpointing with clear comments.
+- Checkpoint state regularly to resume from the last successful sync and cleanly send data to your warehouse.
+- Refer to the [Connector SDK Best Practices](https://fivetran.com/docs/connectors/connector-sdk/best-practices).
+
+---
+
+## 5. Example development guidelines
+
+### 5.1 Example principles
+
+- Self-explanatory and well-commented code.
+- Must produce working code (can require the SDK Playground).
+- No identifiable customer information should be included.
+
+
+
+### 5.2 Demonstrate one core concept
+
+- Each example should focus on one SDK feature or pattern (e.g., incremental sync, schema discovery, error handling).
+- Split complex connectors into multiple examples.
+
+
+
+### 5.3 Avoid overloading
+
+- Don’t demonstrate all features in one example.
+- Create separate folders/examples, each with a distinct goal.
+- Keep the `update()` method short — use smaller helper methods.
+
+
+### 5.4 Use a clear, self-contained folder
+
+- Use descriptive names like `incremental_sync_example` or `dynamic_schema_example`.
+- Include a simple `README.md` describing:
+  - What the example does.
+  - How to run it.
+
+
+
+### 5.5 Minimal dependencies
+
+- Include only necessary dependencies.
+- Pin dependency versions in `requirements.txt`, `pyproject.toml`, or `poetry.lock`.
+
+
+### 5.6 Code organization
+
+- Keep the code easy to read and navigate.
+- Use good naming instead of excessive commenting.
```

Comment:
I don't think it's either or. Good naming and detailed commenting is what works. LongNamesDescribingImplementationDetailsOfIncrementalSyncsForCaseWhereTableHasNoPrimaryKeys does not replace comments, excessive or not.